Suggest annotating traits rather than methods when appropriate

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d2a32fe1f3460b6583534eb53d045165

```rust
fn foo() {
    let _ = "foo".into::<Option<_>>();
}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
   --> src/lib.rs:2:19
    |
2   |     let _ = "foo".into::<Option<_>>();
    |                   ^^^^------------- help: remove these generics
    |                   |
    |                   expected 0 generic arguments
    |
note: associated function defined here, with 0 generic parameters

For more information about this error, try `rustc --explain E0107`.
error: could not compile `playground` due to previous error
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

```
error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
   --> src/lib.rs:2:19
    |
2   |     let _ = "foo".into::<Option<_>>();
    |                   ^^^^------------- help: remove these generics
    |                   |
    |                   expected 0 generic arguments
    |
note: associated function defined here, with 0 generic parameters
help: `Into` trait takes 1 generic argument, try annotating that instead: `Into::<Option<_>>::into("foo")
```

When a user called `into()` with an annotation, it's helpful to suggest `T::from(_)` too.

<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"chansuke"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
@@ -697,7 +697,7 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                         num_trait_generics_except_self
                     )
                 },
-                // TODO(hkmatsumoto): Emit similar suggestion for "x.<assoc fn>()"
+                // FIXME(hkmatsumoto): Emit similar suggestion for "x.<assoc fn>()"
                 hir::ExprKind::MethodCall(..) => return,
                 _ => return,
             }

--- a/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
@@ -683,6 +683,14 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
         let num_trait_generics_except_self =
             trait_generics.count() - if trait_generics.has_self { 1 } else { 0 };
 
+        let msg = format!(
+            "consider moving {these} generic argument{s} to the `{name}` trait, which takes up to {num} argument{s}",
+            these = pluralize!("this", num_assoc_fn_excess_args),
+            s = pluralize!(num_assoc_fn_excess_args),
+            name = self.tcx.item_name(trait_),
+            num = num_trait_generics_except_self,
+        );
+
         if let Some(hir_id) = self.path_segment.hir_id
         && let Some(parent_node) = self.tcx.hir().find_parent_node(hir_id)
         && let Some(parent_node) = self.tcx.hir().find(parent_node)
@@ -691,14 +699,22 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                 hir::ExprKind::Path(ref qpath) => {
                     self.suggest_moving_args_from_assoc_fn_to_trait_for_qualified_path(
                         err,
-                        trait_,
                         qpath,
+                        msg,
                         num_assoc_fn_excess_args,
                         num_trait_generics_except_self
                     )
                 },
-                // FIXME(hkmatsumoto): Emit similar suggestion for "x.<assoc fn>()"
-                hir::ExprKind::MethodCall(..) => return,
+                hir::ExprKind::MethodCall(..) => {
+                    self.suggest_moving_args_from_assoc_fn_to_trait_for_method_call(
+                        err,
+                        trait_,
+                        expr,
+                        msg,
+                        num_assoc_fn_excess_args,
+                        num_trait_generics_except_self
+                    )
+                },
                 _ => return,
             }
         }
@@ -707,8 +723,8 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
     fn suggest_moving_args_from_assoc_fn_to_trait_for_qualified_path(
         &self,
         err: &mut Diagnostic,
-        trait_: DefId,
         qpath: &'tcx hir::QPath<'tcx>,
+        msg: String,
         num_assoc_fn_excess_args: usize,
         num_trait_generics_except_self: usize,
     ) {
@@ -719,13 +735,6 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
             if num_assoc_fn_excess_args == num_trait_generics_except_self - num_generic_args_supplied_to_trait {
                 if let Some(span) = self.gen_args.span_ext()
                 && let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
-                    let msg = format!(
-                        "consider moving {these} generic argument{s} to the `{name}` trait, which takes up to {num} argument{s}",
-                        these = pluralize!("this", num_assoc_fn_excess_args),
-                        s = pluralize!(num_assoc_fn_excess_args),
-                        name = self.tcx.item_name(trait_),
-                        num = num_trait_generics_except_self,
-                    );
                     let sugg = vec![
                         (self.path_segment.ident.span, format!("{}::{}", snippet, self.path_segment.ident)),
                         (span.with_lo(self.path_segment.ident.span.hi()), "".to_owned())
@@ -736,6 +745,28 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                         sugg,
                         Applicability::MaybeIncorrect
                     );
+                }
+            }
+        }
+    }
+
+    fn suggest_moving_args_from_assoc_fn_to_trait_for_method_call(
+        &self,
+        err: &mut Diagnostic,
+        trait_: DefId,
+        expr: &'tcx hir::Expr<'tcx>,
+        msg: String,
+        num_assoc_fn_excess_args: usize,
+        num_trait_generics_except_self: usize,
+    ) {
+        if let hir::ExprKind::MethodCall(_, args, _) = expr.kind {
+            assert_eq!(args.len(), 1);
+            if num_assoc_fn_excess_args == num_trait_generics_except_self {
+                if let Some(gen_args) = self.gen_args.span_ext()
+                && let Ok(gen_args) = self.tcx.sess.source_map().span_to_snippet(gen_args)
+                && let Ok(args) = self.tcx.sess.source_map().span_to_snippet(args[0].span) {
+                    let sugg = format!("{}::{}::{}({})", self.tcx.item_name(trait_), gen_args, self.tcx.item_name(self.def_id), args);
+                    err.span_suggestion(expr.span, msg, sugg, Applicability::MaybeIncorrect);
                 }
             }
         }

--- a/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
@@ -524,8 +524,8 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                 if self.not_enough_args_provided() {
                     self.suggest_adding_args(err);
                 } else if self.too_many_args_provided() {
+                    self.suggest_moving_args_from_assoc_fn_to_trait(err);
                     self.suggest_removing_args_or_generics(err);
-                    self.suggest_moving_args(err);
                 } else {
                     unreachable!();
                 }
@@ -661,34 +661,62 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
     /// ```compile_fail
     /// Into::into::<Option<_>>(42) // suggests considering `Into::<Option<_>>::into(42)`
     /// ```
-    fn suggest_moving_args(&self, err: &mut Diagnostic) {
-        if let Some(trait_) = self.tcx.trait_of_item(self.def_id) {
-            // HACK(hkmatsumoto): Ugly way to tell "<trait>::<assoc fn>()" from "x.<assoc fn>()";
-            // we don't care the latter (for now).
-            if self.path_segment.res == Some(hir::def::Res::Err) {
-                return;
+    fn suggest_moving_args_from_assoc_fn_to_trait(&self, err: &mut Diagnostic) {
+        let trait_ = match self.tcx.trait_of_item(self.def_id) {
+            Some(def_id) => def_id,
+            None => return,
+        };
+
+        // Skip suggestion when the associated function is itself generic, it is unclear
+        // how to split the provided parameters between those to suggest to the trait and
+        // those to remain on the associated type.
+        let num_assoc_fn_expected_args =
+            self.num_expected_type_or_const_args() + self.num_expected_lifetime_args();
+        if num_assoc_fn_expected_args > 0 {
+            return;
+        }
+
+        let num_assoc_fn_excess_args =
+            self.num_excess_type_or_const_args() + self.num_excess_lifetime_args();
+
+        let trait_generics = self.tcx.generics_of(trait_);
+        let num_trait_generics_except_self =
+            trait_generics.count() - if trait_generics.has_self { 1 } else { 0 };
+
+        if let Some(hir_id) = self.path_segment.hir_id
+        && let Some(parent_node) = self.tcx.hir().find_parent_node(hir_id)
+        && let Some(parent_node) = self.tcx.hir().find(parent_node)
+        && let hir::Node::Expr(expr) = parent_node {
+            match expr.kind {
+                hir::ExprKind::Path(ref qpath) => {
+                    self.suggest_moving_args_from_assoc_fn_to_trait_for_qualified_path(
+                        err,
+                        trait_,
+                        qpath,
+                        num_assoc_fn_excess_args,
+                        num_trait_generics_except_self
+                    )
+                },
+                // TODO(hkmatsumoto): Emit similar suggestion for "x.<assoc fn>()"
+                hir::ExprKind::MethodCall(..) => return,
+                _ => return,
             }
+        }
+    }
 
-            // Say, if the assoc fn takes `A`, `B` and `C` as generic arguments while expecting 1
-            // argument, and its trait expects 2 arguments. It is hard to "split" them right as
-            // there are too many cases to handle: `A` `B` | `C`, `A` `B` | `C`, `A` `C` | `B`, ...
-            let num_assoc_fn_expected_args =
-                self.num_expected_type_or_const_args() + self.num_expected_lifetime_args();
-            if num_assoc_fn_expected_args > 0 {
-                return;
-            }
+    fn suggest_moving_args_from_assoc_fn_to_trait_for_qualified_path(
+        &self,
+        err: &mut Diagnostic,
+        trait_: DefId,
+        qpath: &'tcx hir::QPath<'tcx>,
+        num_assoc_fn_excess_args: usize,
+        num_trait_generics_except_self: usize,
+    ) {
+        if let hir::QPath::Resolved(_, path) = qpath
+        && let Some(trait_path_segment) = path.segments.get(0) {
+            let num_generic_args_supplied_to_trait = trait_path_segment.args().num_generic_params();
 
-            let num_assoc_fn_excess_args =
-                self.num_excess_type_or_const_args() + self.num_excess_lifetime_args();
-
-            let trait_generics = self.tcx.generics_of(trait_);
-            let num_trait_generics_except_self =
-                trait_generics.count() - if trait_generics.has_self { 1 } else { 0 };
-
-            // FIXME(hkmatsumoto): RHS of this condition ideally should be
-            // `num_trait_generics_except_self` - "# of generic args already provided to trait"
-            // but unable to get that information with `self.def_id`.
-            if num_assoc_fn_excess_args == num_trait_generics_except_self {
+            if num_assoc_fn_excess_args == num_trait_generics_except_self - num_generic_args_supplied_to_trait {
                 if let Some(span) = self.gen_args.span_ext()
                 && let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
                     let msg = format!(

--- a/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
@@ -525,6 +525,7 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                     self.suggest_adding_args(err);
                 } else if self.too_many_args_provided() {
                     self.suggest_removing_args_or_generics(err);
+                    self.suggest_moving_args(err);
                 } else {
                     unreachable!();
                 }
@@ -650,6 +651,64 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
                 debug!("sugg: {:?}", sugg);
 
                 err.span_suggestion_verbose(sugg_span, &msg, sugg, Applicability::HasPlaceholders);
+            }
+        }
+    }
+
+    /// Suggests moving redundant argument(s) of an associate function to the
+    /// trait it belongs to.
+    ///
+    /// ```compile_fail
+    /// Into::into::<Option<_>>(42) // suggests considering `Into::<Option<_>>::into(42)`
+    /// ```
+    fn suggest_moving_args(&self, err: &mut Diagnostic) {
+        if let Some(trait_) = self.tcx.trait_of_item(self.def_id) {
+            // HACK(hkmatsumoto): Ugly way to tell "<trait>::<assoc fn>()" from "x.<assoc fn>()";
+            // we don't care the latter (for now).
+            if self.path_segment.res == Some(hir::def::Res::Err) {
+                return;
+            }
+
+            // Say, if the assoc fn takes `A`, `B` and `C` as generic arguments while expecting 1
+            // argument, and its trait expects 2 arguments. It is hard to "split" them right as
+            // there are too many cases to handle: `A` `B` | `C`, `A` `B` | `C`, `A` `C` | `B`, ...
+            let num_assoc_fn_expected_args =
+                self.num_expected_type_or_const_args() + self.num_expected_lifetime_args();
+            if num_assoc_fn_expected_args > 0 {
+                return;
+            }
+
+            let num_assoc_fn_excess_args =
+                self.num_excess_type_or_const_args() + self.num_excess_lifetime_args();
+
+            let trait_generics = self.tcx.generics_of(trait_);
+            let num_trait_generics_except_self =
+                trait_generics.count() - if trait_generics.has_self { 1 } else { 0 };
+
+            // FIXME(hkmatsumoto): RHS of this condition ideally should be
+            // `num_trait_generics_except_self` - "# of generic args already provided to trait"
+            // but unable to get that information with `self.def_id`.
+            if num_assoc_fn_excess_args == num_trait_generics_except_self {
+                if let Some(span) = self.gen_args.span_ext()
+                && let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                    let msg = format!(
+                        "consider moving {these} generic argument{s} to the `{name}` trait, which takes up to {num} argument{s}",
+                        these = pluralize!("this", num_assoc_fn_excess_args),
+                        s = pluralize!(num_assoc_fn_excess_args),
+                        name = self.tcx.item_name(trait_),
+                        num = num_trait_generics_except_self,
+                    );
+                    let sugg = vec![
+                        (self.path_segment.ident.span, format!("{}::{}", snippet, self.path_segment.ident)),
+                        (span.with_lo(self.path_segment.ident.span.hi()), "".to_owned())
+                    ];
+
+                    err.multipart_suggestion(
+                        msg,
+                        sugg,
+                        Applicability::MaybeIncorrect
+                    );
+                }
             }
         }
     }

--- a/src/test/ui/const-generics/invalid-const-arg-for-type-param.stderr
+++ b/src/test/ui/const-generics/invalid-const-arg-for-type-param.stderr
@@ -2,15 +2,22 @@ error[E0107]: this associated function takes 0 generic arguments but 1 generic a
   --> $DIR/invalid-const-arg-for-type-param.rs:6:23
    |
 LL |     let _: u32 = 5i32.try_into::<32>().unwrap();
-   |                       ^^^^^^^^------ help: remove these generics
-   |                       |
-   |                       expected 0 generic arguments
+   |                       ^^^^^^^^ expected 0 generic arguments
    |
 note: associated function defined here, with 0 generic parameters
   --> $SRC_DIR/core/src/convert/mod.rs:LL:COL
    |
 LL |     fn try_into(self) -> Result<T, Self::Error>;
    |        ^^^^^^^^
+help: consider moving this generic argument to the `TryInto` trait, which takes up to 1 argument
+   |
+LL |     let _: u32 = TryInto::<32>::try_into(5i32).unwrap();
+   |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+help: remove these generics
+   |
+LL -     let _: u32 = 5i32.try_into::<32>().unwrap();
+LL +     let _: u32 = 5i32.try_into().unwrap();
+   |
 
 error[E0599]: no method named `f` found for struct `S` in the current scope
   --> $DIR/invalid-const-arg-for-type-param.rs:9:7

--- a/src/test/ui/suggestions/issue-89064.rs
+++ b/src/test/ui/suggestions/issue-89064.rs
@@ -24,9 +24,7 @@ fn main() {
     //~| HELP remove these generics
     //~| HELP consider moving these generic arguments
 
-    // bad suggestion
     let _ = A::<S>::foo::<S>();
     //~^ ERROR
     //~| HELP remove these generics
-    //~| HELP consider moving this generic argument
 }

--- a/src/test/ui/suggestions/issue-89064.rs
+++ b/src/test/ui/suggestions/issue-89064.rs
@@ -27,4 +27,9 @@ fn main() {
     let _ = A::<S>::foo::<S>();
     //~^ ERROR
     //~| HELP remove these generics
+
+    let _ = 42.into::<Option<_>>();
+    //~^ ERROR
+    //~| HELP remove these generics
+    //~| HELP consider moving this generic argument
 }

--- a/src/test/ui/suggestions/issue-89064.rs
+++ b/src/test/ui/suggestions/issue-89064.rs
@@ -1,0 +1,32 @@
+use std::convert::TryInto;
+
+trait A<T> {
+    fn foo() {}
+}
+
+trait B<T, U> {
+    fn bar() {}
+}
+
+struct S;
+
+impl<T> A<T> for S {}
+impl<T, U> B<T, U> for S {}
+
+fn main() {
+    let _ = A::foo::<S>();
+    //~^ ERROR
+    //~| HELP remove these generics
+    //~| HELP consider moving this generic argument
+
+    let _ = B::bar::<S, S>();
+    //~^ ERROR
+    //~| HELP remove these generics
+    //~| HELP consider moving these generic arguments
+
+    // bad suggestion
+    let _ = A::<S>::foo::<S>();
+    //~^ ERROR
+    //~| HELP remove these generics
+    //~| HELP consider moving this generic argument
+}

--- a/src/test/ui/suggestions/issue-89064.stderr
+++ b/src/test/ui/suggestions/issue-89064.stderr
@@ -9,15 +9,15 @@ note: associated function defined here, with 0 generic parameters
    |
 LL |     fn foo() {}
    |        ^^^
-help: remove these generics
-   |
-LL -     let _ = A::foo::<S>();
-LL +     let _ = A::foo();
-   |
 help: consider moving this generic argument to the `A` trait, which takes up to 1 argument
    |
 LL -     let _ = A::foo::<S>();
 LL +     let _ = A::<S>::foo();
+   |
+help: remove these generics
+   |
+LL -     let _ = A::foo::<S>();
+LL +     let _ = A::foo();
    |
 
 error[E0107]: this associated function takes 0 generic arguments but 2 generic arguments were supplied
@@ -31,15 +31,15 @@ note: associated function defined here, with 0 generic parameters
    |
 LL |     fn bar() {}
    |        ^^^
-help: remove these generics
-   |
-LL -     let _ = B::bar::<S, S>();
-LL +     let _ = B::bar();
-   |
 help: consider moving these generic arguments to the `B` trait, which takes up to 2 arguments
    |
 LL -     let _ = B::bar::<S, S>();
 LL +     let _ = B::<S, S>::bar();
+   |
+help: remove these generics
+   |
+LL -     let _ = B::bar::<S, S>();
+LL +     let _ = B::bar();
    |
 
 error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
@@ -56,6 +56,27 @@ note: associated function defined here, with 0 generic parameters
 LL |     fn foo() {}
    |        ^^^
 
-error: aborting due to 3 previous errors
+error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/issue-89064.rs:31:16
+   |
+LL |     let _ = 42.into::<Option<_>>();
+   |                ^^^^ expected 0 generic arguments
+   |
+note: associated function defined here, with 0 generic parameters
+  --> $SRC_DIR/core/src/convert/mod.rs:LL:COL
+   |
+LL |     fn into(self) -> T;
+   |        ^^^^
+help: consider moving this generic argument to the `Into` trait, which takes up to 1 argument
+   |
+LL |     let _ = Into::<Option<_>>::into(42);
+   |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+help: remove these generics
+   |
+LL -     let _ = 42.into::<Option<_>>();
+LL +     let _ = 42.into();
+   |
+
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0107`.

--- a/src/test/ui/suggestions/issue-89064.stderr
+++ b/src/test/ui/suggestions/issue-89064.stderr
@@ -43,26 +43,18 @@ LL +     let _ = B::<S, S>::bar();
    |
 
 error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
-  --> $DIR/issue-89064.rs:28:21
+  --> $DIR/issue-89064.rs:27:21
    |
 LL |     let _ = A::<S>::foo::<S>();
-   |                     ^^^ expected 0 generic arguments
+   |                     ^^^----- help: remove these generics
+   |                     |
+   |                     expected 0 generic arguments
    |
 note: associated function defined here, with 0 generic parameters
   --> $DIR/issue-89064.rs:4:8
    |
 LL |     fn foo() {}
    |        ^^^
-help: remove these generics
-   |
-LL -     let _ = A::<S>::foo::<S>();
-LL +     let _ = A::<S>::foo();
-   |
-help: consider moving this generic argument to the `A` trait, which takes up to 1 argument
-   |
-LL -     let _ = A::<S>::foo::<S>();
-LL +     let _ = A::<S>::<S>::foo();
-   |
 
 error: aborting due to 3 previous errors
 

--- a/src/test/ui/suggestions/issue-89064.stderr
+++ b/src/test/ui/suggestions/issue-89064.stderr
@@ -1,0 +1,69 @@
+error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/issue-89064.rs:17:16
+   |
+LL |     let _ = A::foo::<S>();
+   |                ^^^ expected 0 generic arguments
+   |
+note: associated function defined here, with 0 generic parameters
+  --> $DIR/issue-89064.rs:4:8
+   |
+LL |     fn foo() {}
+   |        ^^^
+help: remove these generics
+   |
+LL -     let _ = A::foo::<S>();
+LL +     let _ = A::foo();
+   |
+help: consider moving this generic argument to the `A` trait, which takes up to 1 argument
+   |
+LL -     let _ = A::foo::<S>();
+LL +     let _ = A::<S>::foo();
+   |
+
+error[E0107]: this associated function takes 0 generic arguments but 2 generic arguments were supplied
+  --> $DIR/issue-89064.rs:22:16
+   |
+LL |     let _ = B::bar::<S, S>();
+   |                ^^^ expected 0 generic arguments
+   |
+note: associated function defined here, with 0 generic parameters
+  --> $DIR/issue-89064.rs:8:8
+   |
+LL |     fn bar() {}
+   |        ^^^
+help: remove these generics
+   |
+LL -     let _ = B::bar::<S, S>();
+LL +     let _ = B::bar();
+   |
+help: consider moving these generic arguments to the `B` trait, which takes up to 2 arguments
+   |
+LL -     let _ = B::bar::<S, S>();
+LL +     let _ = B::<S, S>::bar();
+   |
+
+error[E0107]: this associated function takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/issue-89064.rs:28:21
+   |
+LL |     let _ = A::<S>::foo::<S>();
+   |                     ^^^ expected 0 generic arguments
+   |
+note: associated function defined here, with 0 generic parameters
+  --> $DIR/issue-89064.rs:4:8
+   |
+LL |     fn foo() {}
+   |        ^^^
+help: remove these generics
+   |
+LL -     let _ = A::<S>::foo::<S>();
+LL +     let _ = A::<S>::foo();
+   |
+help: consider moving this generic argument to the `A` trait, which takes up to 1 argument
+   |
+LL -     let _ = A::<S>::foo::<S>();
+LL +     let _ = A::<S>::<S>::foo();
+   |
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0107`.

